packages/docusaurus-theme-classic/src/theme/hooks/useCollapse.ts

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import {useState, useEffect, CSSProperties, useRef} from 'react';
+import type {
+  useCollapseProps,
+  useCollapseReturns,
+} from '@theme/hooks/useCollapse';
+
+function getAutoHeightDuration(height) {
+  const constant = height / 36;
+  return Math.round((4 + 15 * constant ** 0.25 + constant / 5) * 10);
```

Comment:
Dynamic transition duration is used in Material deisign, this technique is good for a large number of items. https://material.io/archive/guidelines/motion/duration-easing.html#duration-easing-dynamic-durations